### DISCUSSION TITLE
Breaking changes v2.0.0

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre-gl-js#290] [General]
Version 2.0.0 of this library is coming soon. The migration to Typescript removes support for internet explorer and the public API does not expose the version anymore. See `CHANGELOG.md` of the `typescript` branch.

Breaking changes that most likely will follow are:

* Removal of `accessToken` support
* Removal of the Mapbox CSS classes

**Comment by wipfli** [comment]
While the access token probably has to go, I think it is still up for discussion if we also want to remove the Mapbox CSS classes, see https://github.com/maplibre/maplibre-gl-js/pull/203.

**Comment by wipfli** [comment]
@p-j I am worried that if we remove the Mapbox CSS classes many plugins will break. But then if we keep them the plugins might still break since Mapbox and MapLibre will diverge over time.

**Reply by p-j** [reply]
My 2 cents:
- as long as the general API doesn't change from the last OSS version of mapbox, it's ok to keep the CSS
- as soon as breaking change are introduced in the API of maplibre, then it's ok to start ditching mapbox CSS class names

My preferred course of actions would be:
- release a v2 with a modern / cleaned up stack (TS / CI & CD etc...) keep the CSS and most of the API compatible with v1; clear doc on how to easily switch from mapbox v1 to maplibre v2
- start the work on the v3 that would introduce breaking change and require the ecosystem to adapt; add clear doc on migration, tradeoffs etc...

This way, v1 can be deprecated and upgrade to v2 is a first step toward modern maplibre. V3 is where maplibre clearly diverge from mapbox and start its own path.

A counter point being that it might be too much work for a limited number of maintainers and it might be a better use of resources to keep only v1 as a compatibility layer and deprecate it in favor of a v2 that would start as compatible but quickly break away.

Generally speaking, I think it would be good to invite maintainers of popular compatible libraries to share inputs and insights, either here on GitHub or on a dedicated Slack/Discord.

**Reply by wipfli** [reply]
Cool, thanks. Can you ping some maintainers of popular compatible libraries?

**Reply by HarelM** [reply]
Due to compilation issues related to typescript I needed to break a few APIs. So the typescript branch is 2.0 for sure.
I want to break as many things possible right now to avoid the need to create a version 3 soon.
So accesstoken in on the table and css as well, besides the IE and version API, as mentioned above.
Also since the codebase had change dramatically to support node 14 and the typescript I feel good about saying that it's 2.0 and not 1.x, not just for the outside point of view but also for the developers community...
Either way I see it, typescript branch can't be 1.x. We can discuss if we want 2.x to contain only typecript changes or something else, but I think breaking as much as possible now is the right approach.
I'm not familiar with other things that I would like to break besides the accesstoken and css, so if you feel other stuff need to break now is the time... :-)

**Reply by p-j** [reply]
> Cool, thanks. Can you ping some maintainers of popular compatible libraries?

A first one would be @Pessimistress from [react-map-gl](https://github.com/visgl/react-map-gl) who expressed the possibility of supporting a diverging maplibre in a future version (see [here](https://github.com/visgl/react-map-gl/issues/1513#issuecomment-876822725)) which is cool since `react-map-gl` is one of the go-to React integration.

The other reference integration in React-land is [react-mapbox-gl](https://github.com/alex3165/react-mapbox-gl) by @alex3165 and although the development has slowed, I think it's still quite popular, judging by the number of open PR & issues; but so far [this issue hasn't gotten much attention](https://github.com/alex3165/react-mapbox-gl/issues/924)

Then you have all the addons maintained by mapbox itself:
- [mapbox-gl-geocoder](https://github.com/mapbox/mapbox-gl-geocoder)
- [mapbox-gl-language](https://github.com/mapbox/mapbox-gl-language)
- [mapbox-gl-draw](https://github.com/mapbox/mapbox-gl-draw)
- [mapbox-gl-rtl-text](https://github.com/mapbox/mapbox-gl-rtl-text)

Which I think will need to be forked if they were to become incompatible due to some API/CSS naming change. License might also be an issue here.

If some community members want to ping maintainers or link to projects in other languages (angular, react native, flutter...) to get more inputs and build consensus I think that would be helpful.

**Reply by kylebarron** [reply]
Maybe also ping @stevage of https://github.com/stevage/map-gl-utils

**Reply by stevage** [reply]
Well, hi. My take is that I don't understand the push to make MapLibre incompatible with Mapbox GL, and for such seemingly trivial reasons. Why ditch the CSS classes? They cost almost nothing, and will just randomly break stuff. 

Instead of being able to just easily switch from Mapbox GL JS, now you will have extra migration work, to change your CSS classes. (And if you're using a plugin that relies on those CSS classes, you're probably stuck). IMHO you need a good reason to break stuff, and I don't see it.

Similarly with the accessToken - maybe I'm not understanding it, but this just immediately breaks any attempt to use Mapbox styles, APIs etc? Why do this?

Sometimes breaking changes need to happen, because the cost of not making them is too high. But what it looks like here is a desire to break stuff for the hell of it, to signal some kind of divergence. An approach I am very much opposed to.

**Reply by HarelM** [reply]
We want to create version 2.0 to allow breaking things (and also because the migration to typescript created this situation, although it can be fixed I see this as opportunity). If we are breaking things I believe it would be ideal to do as much as we can in this version to avoid more breaking changes in the future.
As for the reasoning for diverting:
Yes, we want to diverge to avoid cases like #190 the further we diverge the safer this project is.
The name mapbox is a trademark and I would like to remove as much as possible from the code to avoid lawsuits, yes, incuding css. This is probably the main concern here and what drives this from my point of view.
If you want to use a compatible version or continue to support this a branch of 1.x will be available for this. I personally would like to avoid maintaining it, but that just me and other people in the community can decide how to spend their time. :-)
I can't stress how fragile this situation is, at least this is how I perceive it...

**Reply by stevage** [reply]
Thanks for the context. I wouldn't say I'm much closer to  agreeing with the direction here. What I see:
- an urgent push to convert everything to TypeScript (with few direct benefits)
- an urgent push to change the module structure to ESM in order to support Node 12 (again, I don't really understand the benefit)
- breaking changes seen as positive ("the further we diverge the safer this project is") rather than negatives, which is really backwards for most projects

🤷

**Reply by HarelM** [reply]
I probably won't be able convince you, but I respect your opinion. We have a steering committee meeting today and I'll bring it up there.

**Reply by p-j** [reply]
@stevage I agree with you, these are points I've brought forward when discussing the CSS class issue [here](https://github.com/maplibre/maplibre-gl-js/issues/83).
I don't believe using the Mapbox name in code or CSS violates any law or license as long as it's a derivative of the original open source code released by Mapbox .

Switching to TS and renaming objects & CSS won't protect Maplibre from people that disregard policies and copy/paste code from Mapbox v2, it will just make it harder to spot at review time.

However, making a "modern" fork of Mapbox v1 by switching to TS, stopping support for older browsers/node env etc... Is something that make sense as it will reduce the amount of work needed from the maintainers to move forward once that migration is completed. It will also facilitate the onboarding of new contributors as the setup is simplified. But intentionally breaking compatibility for no good reasons beside "branding" is something I don't understand.

In my view, Maplibre should strive to keep compatibility as high as possible and document any breaking changes that had to be made with a migration path. This is the only way the project can attract and retain users migrating from Mapbox v1.

I do not agree with the logic of breaking as many things as possible at once. This will make supporting both Mapbox & Maplibre very difficult for plugin authors, and encourage forks and rewrites, which is not guarantee to happen, will slow down adoption and fragment the ecosystem.

**Reply by HarelM** [reply]
One of the things discussed in #41 is the idea of creating a plugin for mablibre to support this compatibility. I think this idea is a good middle ground - people who are interested in working with both maplibre and mapbox could contribute to this plugin. This will allow this project and repo to continue forward without the need for extra complexity of keeping backwards compatibility which is a big burden on development from my past experience. I think that creating a plugin like this and having the relevant css there and maybe other functions to facilitate smooth transition and compatibility is the right direction. Feel free to let me know if I'm missing something here and if this is not possible in any way.
I would like to focus on developing maplibre and less about maintaining support for mapbox when working on this project... :-)

**Reply by Pessimistress** [reply]
FWIW we are making a push in react-map-gl to be impartial about the map renderer in use. The latest `react-map-gl@beta` release lists mapbox-gl as a peer dependency instead of dependency. We have also removed all hard-coded `mapbox-*` class names from the entire code base. You can test it here: https://github.com/visgl/react-map-gl/tree/7.0-dev/examples/get-started/maplibre

The biggest compatibility risk for our upcoming release is that it heavily relies on `@types/mapbox-gl`, e.g. the react component prop types for Map, Marker etc. extend those that are imported from `@types/mapbox-gl`. If maplibre and mapbox-gl diverge further, TypeScript may start complaining. Please advise if you know of any best practices/solutions of "drop-in replacement" for typing.

**Reply by HarelM** [reply]
@Pessimistress thanks for this wonderful news! From version 2.x we are moving to use the typing straight from the code. This introduce breaking changes in the typings unfortunately, I'm not sure I know how to solve this easily.
You can test and see how this behaves by installing maplibre-gl@2.0.0-pre.6.
While maintaining ngx-maplire-gl I forked the ngx-mapbox-gl and it allowed me to easily support only maplibre, but if you manage to support both maplibre and mapbox using types that are not the same I would love to know how. :-)

**Reply by astridx** [reply]
I have tested https://github.com/visgl/react-map-gl/tree/7.0-dev/examples/get-started/maplibre

At first everything looked good.

~~But when I changed [mapStyle ](https://github.com/visgl/react- good.~~
I wrote too quickly. The other `mapStyle` were also displayed after I called `npm run build` again.

**Reply by joshuacrew** [reply]
> I have tested https://github.com/visgl/react-map-gl/tree/7.0-dev/examples/get-started/maplibre

I found an issue with combining that example with the Controls example [here](https://github.com/visgl/react-map-gl/tree/7.0-dev/examples/controls). I have put an [issue](https://github.com/visgl/react-map-gl/issues/1685) in the react-map-gl library, but it may be more applicable here.

**Comment by wipfli** [comment]
Updates from the steering committee meeting: MapLibre's priority is to make it easy for users to switch from Mapbox v1 to MapLibre v2. The release notes should document what steps are needed for such a switch. Avoiding fragmentation has beed and will continue to be of highest importance to this project and introducing breaking changes for symbolic reasons are not acceptable. Maintainers of this code base want to focus on new features and improvements. Keeping it compatible with Mapbox v2 is however not required.

**Reply by wipfli** [reply]
@marcelnormann @nyurik @HarelM @klokan were in today's meeting. Correct me if I misinterpreted something.

**Comment by stevage** [comment]
Thanks for the update, much appreciated.

**Comment by arichnad** [comment]
I've read the v2 pre release changelog, and it says `** Breaking Change ** removed all code related to accessToken and mapbox specific urls`, but it doesn't say what to do about this?  Having a one-liner about what needs to change or even a pointer on what needs to happen will be insanely useful to users upgrading to v2 or anyone trying to change from mapbox to maplibre.

Maybe something along the lines of, change `"mapbox://...")` to `"style.json"` where style.json has something specific in it?

> The release notes should document what steps are needed for such a switch

@wipfli Where are these release notes?  I don't see them here or in the maplibre-gl-js-docs.  Linking from the changelog to this document might be useful, thanks!

**Reply by HarelM** [reply]
This is a community project, feel free to submit a PR with this relevant changes if you think the current description is lacking.
I have read through the licence note in the removed files related to mapbox usage of the accesstoken and I can say that I'm not willing to risk a law suite related to "documentation on how to use mapbox illegally" personally. The bottom line is: don't use mapbox with this library. If this is clear enough, as I said, feel free to submit a PR. :-)

**Reply by wipfli** [reply]
Thanks for joining the discussion @arichnad. Indeed as Harel said it is a really delicate topic for MapLibre. Would you like to use MapLibre GL JS with Mapbox sources? If this is the case, what is better here than with Mapbox v2?

**Reply by arichnad** [reply]
> documentation on how to use mapbox illegally

Ha, sorry I didn't realize that's what I was doing!

> feel free to submit a PR

Uhh, ok?  I don't know your process at all, sorry, but . . . #336 ???

> Would you like to use MapLibre GL JS with Mapbox sources?

Yes and no.  I use both an openmaptiles/tileserver-gl mode (primary), and a mapbox satellite mode (secondary).  I very much appreciate that maplibre is open source, and want to continue to work with maplibre, but also sad that mapbox won't allow maplibre to integrate with their content?  I'm not sure how I will proceed, but thank you for responding.

**Reply by wipfli** [reply]
If you are interested in contributing to MapLibre GL JS then you are more than welcome! Actually engaging in this discussion is already very valuable.

**Reply by kylebarron** [reply]
> sad that mapbox won't allow maplibre to integrate with their content?

Do we have any evidence at all that Mapbox doesn't want Maplibre to use Mapbox servers? Correct me if I'm wrong, but Mapbox servers are a public REST API, accessible broadly with an access token. Are there any legal terms anywhere that even suggest that Mapbox servers cannot be used with third party libraries?

It seems quite valuable to end users to be able to mix and match private styles and data with Mapbox styles and data, just as @arichnad was interested in doing.

**Reply by stevage** [reply]
>Correct me if I'm wrong, but Mapbox servers are a public REST API, accessible broadly with an access token. 

My understanding after a brief re-read:
- there is a [vector tiles API](https://docs.mapbox.com/api/maps/vector-tiles/#vector-tiles-api-restrictions-and-limits) which works like this, and is [priced by tile request](https://www.mapbox.com/pricing/#vectortile).
- there is Mapbox GL JS v2 which has additional terms (not relevant here) and [charges per map load](https://www.mapbox.com/pricing/#maploads)


I don't see any indication that there is anything problematic about using a third-party library (which MapLibre essentially is) to access Mapbox services, provided you have an API key etc.

Now, HarelM referenced [the warning in this code](https://github.com/maplibre/maplibre-gl-js/blob/1.x/src/util/mapbox.js), which I'll quote:

```
/***** START WARNING - IF YOU USE THIS CODE WITH MAPBOX MAPPING APIS, REMOVAL OR
* MODIFICATION OF THE FOLLOWING CODE VIOLATES THE MAPBOX TERMS OF SERVICE  ******
* The following code is used to access Mapbox's Mapping APIs. Removal or modification
* of this code when used with Mapbox's Mapping APIs can result in higher fees and/or
* termination of your account with Mapbox.
*
* Under the Mapbox Terms of Service, you may not use this code to access Mapbox
* Mapping APIs other than through Mapbox SDKs.
*
* The Mapping APIs documentation is available at https://docs.mapbox.com/api/maps/#maps
* and the Mapbox Terms of Service are available at https://www.mapbox.com/tos/
******************************************************************************/
```

I read this "warning" as what it is, a warning that modifying code here might be breaking terms defined elsewhere. It is not, itself, a licence or agreement. And the problem is, the Mapbox terms of service keep changing, so it's very hard to know what the terms were in effect at the time this was written, or even to know why it was written.

In particular, I can't find any justification for this statement in the current Mapbox Terms of Service:

>Under the Mapbox Terms of Service, you may not use this code to access Mapbox Mapping APIs other than through Mapbox SDKs.

There isn't anything in the terms at all about SDKs. I'm not certain, but I suspect there were previously some much more restrictive terms that have been removed. Maybe one of these changes:

> April 24, 2020: product terms removed and replaced with a reference to our standard service terms.

> July 1, 2019: Removed the default restriction that Services have to be used in conjunction with a Mapbox map. Point of interest results from the Geocoding API and Enterprise Boundaries must still be used in conjunction with a Mapbox map.

> May 22, 2019: Updates to reflect pricing changes; removal of commercial and private application restrictions; addition of late payment interest and attorneys’ fees; addition of audit right for non-geocoding services; updated tax language; and minor corrections and reorganizations for clarity.


So, in summary: I think it's most likely fine to access Mapbox services through MapLibre, although I don't entirely understand exactly what part of that file is needed anyway (since most of it seems to be about implementing the "Map loads" pricing which isn't relevant here).

And IMHO this:

> The bottom line is: don't use mapbox with this library. 

Is a very disappointing stance for a project maintainer to be taking. I really hope this does not reflect the direction of the project, because if it does, it will make it much less valuable.

**Reply by Pessimistress** [reply]
As of today the Mapbox hosted tiles can still be loaded by appending the access token as a query parameter. Users of maplibre can use a map style that references Mapbox servers in tandem with the `transformRequest` callback:

https://github.com/maplibre/maplibre-gl-js/blob/ac88200b868974f62a44f69f1b4514a98f2ccddf/src/ui/map.ts#L245

I personally think it's perfectly fine for maplibre to include a code sample for this use case in its documentation, alongside a "read the Mapbox T&C before use" warning.

**Reply by wipfli** [reply]
Thanks everyone for the feedback. I am happy to review pulll requests which update the readme or changelog to clarify how one is supposed to use Mapbox services with MapLibre GL JS. All contributions are welcome!

**Reply by igalgh** [reply]
Has anyone been able to use the `RequestTransformFunction` approach to load a mapbox style?
If so, would love to see a sample

**Reply by AbelVM** [reply]
I'm quite interested in this issue too :sweat_smile:

**Comment by nyurik** [comment]
My take on licensing (IANAL) is that this is outside of the MapLibre's scope to evaluate constantly changing terms of service of any data provider, let alone endorse or give any legal guarantees -- from BSD license -- `THIS SOFTWARE IS PROVIDED ... "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES ... ARE DISCLAIMED`.

That said, we always want to "play nice" and provide as much help and transparency as possible. MapBox usage is clearly a goal of many users, so we should have a dedicated section in our documentation/examples/....  The docs should:
* link to MB license/TOS
* document the removal of the mapbox-specific code and associated [scary text](https://github.com/maplibre/maplibre-gl-js/blob/1.x/src/util/mapbox.js), and explain the reasoning behind code removal
* reiterate that MapLibre libs are provider-neutral, and cannot endorse any data provider or do legal evaluation of their terms/give any legal guarantees
* provide MapLibre usage examples based on Mapbox API official docs, i.e. if the docs say that a token must be appended to a tile URL, show how to do it with MapLibre

We should also kindly ask Mapbox to clarify their position on 3rd party usage of their service by non-Mapbox authorized client libraries, including the seeming discrepancy with the "scary text" warning. Perhaps this should be done by a Mapbox customer who wants to use MapLibre while paying for Mapbox service.

P.S. Thanks everyone who is participating!

**Reply by stevage** [reply]
That seems very reasonable.

I have reached out to someone at Mapbox by email. Will keep you updated.

**Reply by sabres207** [reply]
So is it possible to use mapbox api with maplibre even with version 2?
I'm currently using maplibre version 1 only for that reason.

**Reply by aaguilardm** [reply]
Hi, has there been any update from Mapbox in regards to this?

**Reply by stevage** [reply]
No, I haven't heard any further word from them.

**Comment by acalcutt** [comment]
I was testing the MapLibre Terrain3d branch and I noticed it no longer worked with the maplibre-gl-inspect.min.js plugin I had modified from mapbox-gl-inspect.min.js.

After looking into it I think it is related to the removal of 'version', which the plugin used to make sure the library was a supported version. I was able to get it working again by removing that check
var e, n = null;
```
                window.maplibregl ? ((e = (e = window.maplibregl).version.split(".").map(parseFloat))[0] < 1 && e[1] < 29 && console.error("MaplibreInspect only supports Maplibre GL JS >= v0.29.0. Please upgrade your Maplibre GL JS version."), n = new window.maplibregl.Popup({
                    closeButton: !1,
                    closeOnClick: !1
                })
```
was changed to 
```
                window.maplibregl ? ((n = new window.maplibregl.Popup({
                    closeButton: !1,
                    closeOnClick: !1
                }))
```

This should work for my use, but I'm wondering if 'version' is used in other plugins and/or still useful

**Comment by HarelM** [comment]
It's worth mentioning here that we are looking into named exports as part of this version"s breaking changes. The advantage is to allow tree shaking, the disadvantage is that it will split the ecosystem sooner than later.
See #623 for more technical info.
Thoughts are welcome :-)

**Reply by stevage** [reply]
This does not seem like a big deal to me. If I'm understanding right, it basically means changing:

> `import mapboxgl from 'mapbox-gl';`

to

> `import * as mapboxgl from 'maplibre-gl';`

or something similar.

Is it a bigger deal than that? This does not seem like something that "splits the ecosystem" to me.

**Reply by HarelM** [reply]
I'm not sure yet, it might be just that, and in this case, as you said, it's not that big of a deal, but it might be more complicated, I still need to research it and see what are the expected changes.
It might also be a candidate for 3.x in case we decide to release 2.x without it.
The more I read about it the more I'm inclined to change the output format from UMD to ESM, which might not be fully related to named exports but seems somehow connected... I'm not sure...

**Comment by astridx** [comment]
> While the access token probably has to go, I think it is still up for discussion if we also want to remove the Mapbox CSS classes, see #203.

I'm not sure, but didn't we want to delete the classes in version 2? 
If we don't do this, it's only possible for the next major version (3.0), right?
